### ISSUE TITLE
Introduce `unnecessary_function_argument()` linter

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#1755]
In multi-parameter functions, it often happens that some parameters are factored out of the code but are never removed from the function signature. 

A simple example of this would be a function like the following, where the parameter `arg1` is no longer needed because whatever relied upon it in the function body has already been removed, but the author forgot to update the signature.

``` r
library(lintr)

lint(
  text = "f <- function(x, arg1 = TRUE) mean(x)", 
  linters = linters_with_tags(NULL)
)
```

<sup>Created on 2022-10-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

This can be detected by checking if a symbol corresponding to this argument appears anywhere in the function body. Note that this logic (correctly) won't produce lints for deprecated parameters.

Exceptions should be made for 

- `S3` generics

```r
foo <- function(x, ...) {
  UseMethod("foo")
}
```

- `...`, which can sometimes be used to change function signature in a backward compatible way

```r
foo <- function(x, y1 = NULL, ...) { g(x, y1) } # old
foo <- function(x, y2 = NULL, ...) { g(x, y2) } # new; won't break, `y1 = ` will be absorbed by `...`
```

**Comment by IndrajeetPatil** [comment]
This is definitely out there, but we can also try to capture recursive functions with unnecessary argument.

E.g., `y` symbol is encountered in the function body, but that argument is still not used anywhere.

```r
foo <- function(x, y, ...) foo(g(x), y, ...)
```

**Comment by AshesITR** [comment]
That sounds very difficult for edge cases.

```r
foo <- function(x, y, z) {
  if (x == 0) {
    y
  } else {
    (x - 1) %>% foo(z, y)
  }
}
```

That function has no unnecessary arguments but it will probably be very hard to statically conclude this.

**Comment by Bisaloo** [comment]
Would cases where the argument is overwritten by a constant value before being used fit in the scope of this linter?

```r
foo <- function(x) {

  x <- 2
  return(x^2)

}
```

**Comment by MichaelChirico** [comment]
I believe I pursued a linter like this at some point but gave up, the edge cases were hard to tamp down & quite prevalent in a large enough code base.

Not necessarily saying that it's not worth pursuing, but definitely test any potential candidate against a large number of CRAN packages upfront.

**Comment by AshesITR** [comment]
Another edge case comes to mind with glue:

```r
function(x) {
  glue::glue("x = {x}")
}
```

**Comment by aramirezreyes** [comment]
+1 More vote for this kind of linter.
